Upload: interruptUpload throws AssertionError

Vaadin Framework version: 8.8.5
Browser: any
Web container: any

Caused by the fix for #11616
Related issues: #4212 #9635

Sample project to reproduce the issue: https://github.com/willemv/vaadin-upload-bug

To reproduce:
* start a Vaadin-based application with assertions enabled (i.e. by adding the `-ea` JVM parameter)
* create an `Upload` button
* start uploading a file
* call the `interruptUpload` method on the upload button

Result: an `AssertionError` is thrown, because the `FileUploadHandler` tries to access the connector state outside of the session lock:

```
Jul 31, 2019 3:50:17 PM com.vaadin.server.DefaultErrorHandler doDefault
SEVERE: 
java.lang.AssertionError: Session must be locked when getState() is called
	at com.vaadin.server.AbstractClientConnector.getState(AbstractClientConnector.java:264)
	at com.vaadin.ui.AbstractComponent.getState(AbstractComponent.java:676)
	at com.vaadin.ui.AbstractSingleComponentContainer.getState(AbstractSingleComponentContainer.java:364)
	at com.vaadin.ui.UI.getState(UI.java:340)
	at com.vaadin.ui.PushConfigurationImpl.getState(PushConfiguration.java:336)
	at com.vaadin.ui.PushConfigurationImpl.getPushMode(PushConfiguration.java:181)
	at com.vaadin.server.communication.FileUploadHandler.streamToReceiver(FileUploadHandler.java:621)
	at com.vaadin.server.communication.FileUploadHandler.handleFileUploadValidationAndData(FileUploadHandler.java:460)
	at com.vaadin.server.communication.FileUploadHandler.doHandleSimpleMultipartFileUpload(FileUploadHandler.java:413)
	at com.vaadin.server.communication.FileUploadHandler.handleRequest(FileUploadHandler.java:290)
	at com.vaadin.server.VaadinService.handleRequest(VaadinService.java:1602)
	at com.vaadin.server.VaadinServlet.service(VaadinServlet.java:445)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:841)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:535)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:168)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:166)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1155)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)
	at org.eclipse.jetty.server.Server.handle(Server.java:561)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:334)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:104)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:247)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:140)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:243)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:679)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:597)
	at java.lang.Thread.run(Thread.java:748)
```
If you run _without_ assertions, the cancel simply doesn't work

--- a/server/src/main/java/com/vaadin/server/communication/FileUploadHandler.java
+++ b/server/src/main/java/com/vaadin/server/communication/FileUploadHandler.java
@@ -621,11 +621,7 @@ public class FileUploadHandler implements RequestHandler {
             } finally {
                 session.unlock();
             }
-            boolean pushEnabled = UI.getCurrent().getPushConfiguration()
-                    .getPushMode().isEnabled();
-            if (!pushEnabled) {
-                return true;
-            }
+            return true;
 
             // Note, we are not throwing interrupted exception forward as it is
             // not a terminal level error like all other exception.
@@ -704,7 +700,20 @@ public class FileUploadHandler implements RequestHandler {
 
     private void cleanStreamVariable(VaadinSession session, final UI ui,
             final ClientConnector owner, final String variableName) {
-        session.accessSynchronously(() -> ui.getConnectorTracker()
-                .cleanStreamVariable(owner.getConnectorId(), variableName));
+        session.accessSynchronously(() -> {
+            ui.getConnectorTracker().cleanStreamVariable(owner.getConnectorId(),
+                    variableName);
+
+            // in case of automatic push mode, the client connector
+            // could already have refreshed its StreamVariable
+            // in the ConnectorTracker. For instance, the Upload component
+            // adds its stream variable in its paintContent method, which is
+            // called (indirectly) on each session unlock in case of automatic
+            // pushes.
+            // To cover this case, mark the client connector as dirty, so that
+            // the unlock after this runnable refreshes the StreamVariable
+            // again.
+            owner.markAsDirty();
+        });
     }
 }

--- a/uitest/src/main/java/com/vaadin/tests/components/upload/InterruptUpload.java
+++ b/uitest/src/main/java/com/vaadin/tests/components/upload/InterruptUpload.java
@@ -22,6 +22,15 @@ public class InterruptUpload extends AbstractTestUI {
 
     private Upload sample;
     private UploadInfoWindow uploadInfoWindow;
+    private final boolean pushManually;
+
+    public InterruptUpload() {
+        this(false);
+    }
+
+    public InterruptUpload(boolean pushManually) {
+        this.pushManually = pushManually;
+    }
 
     @Override
     protected void setup(VaadinRequest request) {
@@ -39,13 +48,17 @@ public class InterruptUpload extends AbstractTestUI {
                 UI.getCurrent().addWindow(uploadInfoWindow);
             }
             uploadInfoWindow.setClosable(false);
+            pushIfManual();
         });
-        sample.addFinishedListener(event -> uploadInfoWindow.setClosable(true));
+        sample.addFinishedListener(event -> {
+            uploadInfoWindow.setClosable(true);
+            pushIfManual();
+        });
 
-        addComponent(sample);
+        addComponents(new Label(getDescription()), sample);
     }
 
-    private static class UploadInfoWindow extends Window
+    private class UploadInfoWindow extends Window
             implements Upload.StartedListener, Upload.ProgressListener,
             Upload.FailedListener, Upload.SucceededListener,
             Upload.FinishedListener {
@@ -114,6 +127,7 @@ public class InterruptUpload extends AbstractTestUI {
             textualProgress.setVisible(false);
             cancelButton.setVisible(false);
             UI.getCurrent().setPollInterval(-1);
+            pushIfManual();
         }
 
         @Override
@@ -128,6 +142,7 @@ public class InterruptUpload extends AbstractTestUI {
             fileName.setValue(event.getFilename());
 
             cancelButton.setVisible(true);
+            pushIfManual();
         }
 
         @Override
@@ -138,11 +153,13 @@ public class InterruptUpload extends AbstractTestUI {
             textualProgress.setValue(
                     "Processed " + readBytes + " bytes of " + contentLength);
             result.setValue(counter.getLineBreakCount() + " (counting...)");
+            pushIfManual();
         }
 
         @Override
         public void uploadSucceeded(final SucceededEvent event) {
             result.setValue(counter.getLineBreakCount() + " (total)");
+            pushIfManual();
         }
 
         @Override
@@ -150,6 +167,13 @@ public class InterruptUpload extends AbstractTestUI {
             result.setValue(
                     counter.getLineBreakCount() + " (counting interrupted at "
                             + Math.round(100 * progressBar.getValue()) + "%)");
+            pushIfManual();
+        }
+    }
+
+    private void pushIfManual() {
+        if (pushManually) {
+            push();
         }
     }
 

--- a/uitest/src/main/java/com/vaadin/tests/components/upload/InterruptUploadWithManualPush.java
+++ b/uitest/src/main/java/com/vaadin/tests/components/upload/InterruptUploadWithManualPush.java
@@ -1,0 +1,23 @@
+package com.vaadin.tests.components.upload;
+
+import com.vaadin.annotations.Push;
+import com.vaadin.shared.communication.PushMode;
+
+@Push(PushMode.MANUAL)
+public class InterruptUploadWithManualPush extends InterruptUpload {
+
+    public InterruptUploadWithManualPush() {
+        super(true);
+    }
+
+    @Override
+    protected Integer getTicketNumber() {
+        return 11616;
+    }
+
+    @Override
+    public String getDescription() {
+        return "Interrupting an upload with @Push shouldn't prevent uploading that same file immediately afterwards.";
+    }
+
+}

--- a/uitest/src/main/java/com/vaadin/tests/dnd/Html5FileDragAndDropUpload.java
+++ b/uitest/src/main/java/com/vaadin/tests/dnd/Html5FileDragAndDropUpload.java
@@ -4,15 +4,18 @@ import java.io.IOException;
 import java.io.OutputStream;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 import com.vaadin.server.StreamVariable;
 import com.vaadin.server.VaadinRequest;
 import com.vaadin.shared.ui.dnd.FileParameters;
 import com.vaadin.shared.ui.grid.DropMode;
 import com.vaadin.tests.components.AbstractTestUIWithLog;
+import com.vaadin.ui.Button;
 import com.vaadin.ui.Grid;
-import com.vaadin.ui.Layout;
 import com.vaadin.ui.Notification;
+import com.vaadin.ui.ProgressBar;
+import com.vaadin.ui.UI;
 import com.vaadin.ui.VerticalLayout;
 import com.vaadin.ui.components.grid.GridDropTarget;
 import com.vaadin.ui.dnd.FileDropTarget;
@@ -20,6 +23,16 @@ import com.vaadin.ui.dnd.FileDropTarget;
 public class Html5FileDragAndDropUpload extends AbstractTestUIWithLog {
 
     private static final int FILE_SIZE_LIMIT = 1024 * 1024 * 5; // 5 MB
+
+    private final boolean pushManually;
+
+    public Html5FileDragAndDropUpload() {
+        this(false);
+    }
+
+    public Html5FileDragAndDropUpload(boolean pushManually) {
+        this.pushManually = pushManually;
+    }
 
     @Override
     protected void setup(VaadinRequest request) {
@@ -31,6 +44,9 @@ public class Html5FileDragAndDropUpload extends AbstractTestUIWithLog {
         grid.addColumn(FileParameters::getMime).setCaption("Mime type");
 
         List<FileParameters> gridItems = new ArrayList<>();
+        AtomicBoolean cancelled = new AtomicBoolean(false);
+
+        ProgressBar progressBar = new ProgressBar(0.0f);
 
         new FileDropTarget<Grid<FileParameters>>(grid,
                 event -> event.getFiles().forEach(html5File -> {
@@ -39,6 +55,7 @@ public class Html5FileDragAndDropUpload extends AbstractTestUIWithLog {
                                 + " is too large (max 5 MB)");
                         return;
                     }
+                    UI.getCurrent().setPollInterval(200);
 
                     html5File.setStreamVariable(new StreamVariable() {
                         @Override
@@ -58,19 +75,27 @@ public class Html5FileDragAndDropUpload extends AbstractTestUIWithLog {
 
                         @Override
                         public void onProgress(StreamingProgressEvent event) {
+                            float progress = (float) event.getBytesReceived()
+                                    / event.getContentLength();
+                            progressBar.setValue(progress);
                             log("Progress, bytesReceived="
                                     + event.getBytesReceived());
+                            pushIfManual();
                         }
 
                         @Override
                         public void streamingStarted(
                                 StreamingStartEvent event) {
+                            cancelled.set(false);
+                            progressBar.setValue(0.0f);
                             log("Stream started, fileName="
                                     + event.getFileName());
+                            pushIfManual();
                         }
 
                         @Override
                         public void streamingFinished(StreamingEndEvent event) {
+                            progressBar.setValue(1.0f);
                             gridItems
                                     .add(new FileParameters(event.getFileName(),
                                             event.getContentLength(),
@@ -79,17 +104,22 @@ public class Html5FileDragAndDropUpload extends AbstractTestUIWithLog {
 
                             log("Stream finished, fileName="
                                     + event.getFileName());
+                            pushIfManual();
+                            UI.getCurrent().setPollInterval(-1);
                         }
 
                         @Override
                         public void streamingFailed(StreamingErrorEvent event) {
+                            progressBar.setValue(0.0f);
                             log("Stream failed, fileName="
                                     + event.getFileName());
+                            pushIfManual();
+                            UI.getCurrent().setPollInterval(-1);
                         }
 
                         @Override
                         public boolean isInterrupted() {
-                            return false;
+                            return cancelled.get();
                         }
                     });
                 }));
@@ -101,9 +131,18 @@ public class Html5FileDragAndDropUpload extends AbstractTestUIWithLog {
             Notification.show(event.getDataTransferText());
         });
 
-        Layout layout = new VerticalLayout(grid);
+        Button cancelButton = new Button("Cancel",
+                click -> cancelled.set(true));
+        VerticalLayout layout = new VerticalLayout(grid, cancelButton,
+                progressBar);
 
         addComponent(layout);
+    }
+
+    private void pushIfManual() {
+        if (pushManually) {
+            push();
+        }
     }
 
     @Override

--- a/uitest/src/main/java/com/vaadin/tests/dnd/Html5FileDragAndDropUploadWithManualPush.java
+++ b/uitest/src/main/java/com/vaadin/tests/dnd/Html5FileDragAndDropUploadWithManualPush.java
@@ -1,0 +1,12 @@
+package com.vaadin.tests.dnd;
+
+import com.vaadin.annotations.Push;
+import com.vaadin.shared.communication.PushMode;
+
+@Push(PushMode.MANUAL)
+public class Html5FileDragAndDropUploadWithManualPush
+        extends Html5FileDragAndDropUpload {
+    public Html5FileDragAndDropUploadWithManualPush() {
+        super(true);
+    }
+}

--- a/uitest/src/main/java/com/vaadin/tests/dnd/Html5FileDragAndDropUploadWithPush.java
+++ b/uitest/src/main/java/com/vaadin/tests/dnd/Html5FileDragAndDropUploadWithPush.java
@@ -1,0 +1,8 @@
+package com.vaadin.tests.dnd;
+
+import com.vaadin.annotations.Push;
+
+@Push
+public class Html5FileDragAndDropUploadWithPush
+        extends Html5FileDragAndDropUpload {
+}

